gr-qtgui/lib/fosphor_formatter_impl.cc

```diff
@@ -12,205 +13,223 @@
 #endif
 
 #include "fosphor_formatter_impl.h"
+#include <gnuradio/fft/fft.h>
 #include <gnuradio/fft/window.h>
 #include <gnuradio/io_signature.h>
+#include <volk/volk_32fc_32f_multiply_32fc.h>
+#include <volk/volk_32fc_x2_multiply_32fc.h>
 #include <cmath>
 #include <cstring>
 
-using namespace gr::qtgui;
+using gr::fft::window;
 
-namespace {
+namespace gr {
+namespace qtgui {
 
-// Helper function for converting float to unsigned char
-void float_array_to_uchar(const float* in,
-                          unsigned char* out,
-                          int nsamples,
-                          const float scaling = 1.0f)
+fosphor_formatter_impl::fosphor_formatter_impl(int fft_size,
+                                               int num_bins,
+                                               int input_decim,
+                                               int waterfall_decim,
+                                               int histo_decim,
+                                               double scale,
+                                               double alpha,
+                                               double epsilon,
+                                               double trise,
+                                               double tdecay)
+    : block("histo_proc",
+            gr::io_signature::make(1, 1, sizeof(gr_complex)),
+            gr::io_signature::make(2, 2, sizeof(unsigned char) * fft_size)),
+      d_fft_size(fft_size),
+      d_num_bins(num_bins),
+      d_input_decim(input_decim),
+      d_waterfall_decim(waterfall_decim),
+      d_histo_decim(histo_decim),
+      d_scale(static_cast<float>(scale)),
+      d_alpha(static_cast<float>(alpha)),
+      d_epsilon(static_cast<float>(epsilon)),
+      d_trise(static_cast<float>(trise)),
+      d_tdecay(static_cast<float>(tdecay)),
+      d_iir(fft_size, 0.0f),
+      d_maxhold_buf(fft_size, 0.0f),
+      d_histo_buf_f(fft_size * d_num_bins, 0.0f),
+      d_histo_buf_b(fft_size * d_num_bins, 0),
+      d_hit_count(fft_size * d_num_bins, 0),
+      d_input_count(0),
+      d_waterfall_count(0),
+      d_histo_count(0)
 {
-    constexpr int MIN_UCHAR = 0;
-    constexpr int MAX_UCHAR = 255;
-    for (int i = 0; i < nsamples; i++) {
-        long int r = (long int)rint(in[i] * scaling);
-        if (r < MIN_UCHAR) {
-            r = MIN_UCHAR;
-        } else if (r > MAX_UCHAR) {
-            r = MAX_UCHAR;
-        }
-        out[i] = r;
-    }
+    // Output is d_num_bins histogram levels, plus average and max rows. Ensure
+    // there is space, in case output is generated.
+    set_min_output_buffer(2 * (d_num_bins + 2));
+
+    // TODO: make sure decims are >0
+
+    d_fft = new gr::fft::fft_complex_fwd(fft_size);
```

Comment:
Shouldn't we put this into a unique_ptr?

Reply:
Yes!